File: CS/HowToBindCommands/CommandUIBindings/Adapters/CommandButtonAdapter.cs
using System;
using System.Windows.Controls;
using DevExpress.Utils.Commands;
using DevExpress.XtraRichEdit;
using HowToBindCommands.CommandUIBindings.UIState;

namespace HowToBindCommands.Adapters
{
    #region #commandbuttonadapter
    public class CommandButtonAdapter
    {
        Command command;
        RichEditControl control;
        Button button;

        public RichEditControl RichEditControl
        {
            get { return control; }
            set
            {
                if (control == value)
                    return;
                UnsubscribeControlEvents();
                this.control = value;
                SubscribeControlEvents();
                OnUpdateUI(this, EventArgs.Empty);
            }
        }
        public Command Command
        {
            get { return command; }
            set
            {
                if (command == value)
                    return;
                command = value;
                OnUpdateUI(this, EventArgs.Empty);
            }
        }

        public Button Button
        {
            get { return button; }
            set
            {
                if (button == value)
                    return;

                UnsubscribeButtonEvents();
                button = value;
                SubscribeButtonEvents();
                OnUpdateUI(this, EventArgs.Empty);
            }
        }
        void SubscribeButtonEvents()
        {
            if (button == null)
                return;

            button.Click += OnClick;
        }
        void UnsubscribeButtonEvents()
        {
            if (button == null)
                return;

            button.Click -= OnClick;
        }
        void SubscribeControlEvents()
        {
            if (control == null)
                return;
            control.UpdateUI += OnUpdateUI;
        }
        void UnsubscribeControlEvents()
        {
            if (control == null)
                return;
            control.UpdateUI -= OnUpdateUI;
        }

        void OnUpdateUI(object sender, EventArgs e)
        {
            if (button == null)
                return;

            Command command = CreateCommand();
            if (command != null)
            {
                CommandButtonUIState state = new CommandButtonUIState(button);
                command.UpdateUIState(state);
            }
        }

        void OnClick(object sender, EventArgs e)
        {
            Command command = CreateCommand();
            if (command != null)
                command.Execute();
        }

        // You may override this method to create a command
        protected virtual Command CreateCommand()
        {
            return command;
        }
    }
    #endregion #commandbuttonadapter
}

File: VB/HowToBindCommands/CommandUIBindings/Adapters/CommandButtonAdapter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Windows.Controls
Imports DevExpress.Utils.Commands
Imports DevExpress.XtraRichEdit
Imports HowToBindCommands.CommandUIBindings.UIState

Namespace HowToBindCommands.Adapters
	#Region "#commandbuttonadapter"
	Public Class CommandButtonAdapter
		Private command_Renamed As Command
		Private control As RichEditControl
		Private button_Renamed As Button

		Public Property RichEditControl() As RichEditControl
			Get
				Return control
			End Get
			Set(ByVal value As RichEditControl)
				If control Is value Then
					Return
				End If
				UnsubscribeControlEvents()
				Me.control = value
				SubscribeControlEvents()
				OnUpdateUI(Me, EventArgs.Empty)
			End Set
		End Property
		Public Property Command() As Command
			Get
				Return command_Renamed
			End Get
			Set(ByVal value As Command)
				If command_Renamed Is value Then
					Return
				End If
				command_Renamed = value
				OnUpdateUI(Me, EventArgs.Empty)
			End Set
		End Property

		Public Property Button() As Button
			Get
				Return button_Renamed
			End Get
			Set(ByVal value As Button)
				If button_Renamed Is value Then
					Return
				End If

				UnsubscribeButtonEvents()
				button_Renamed = value
				SubscribeButtonEvents()
				OnUpdateUI(Me, EventArgs.Empty)
			End Set
		End Property
		Private Sub SubscribeButtonEvents()
			If button_Renamed Is Nothing Then
				Return
			End If

			AddHandler button_Renamed.Click, AddressOf OnClick
		End Sub
		Private Sub UnsubscribeButtonEvents()
			If button_Renamed Is Nothing Then
				Return
			End If

			RemoveHandler button_Renamed.Click, AddressOf OnClick
		End Sub
		Private Sub SubscribeControlEvents()
			If control Is Nothing Then
				Return
			End If
			AddHandler control.UpdateUI, AddressOf OnUpdateUI
		End Sub
		Private Sub UnsubscribeControlEvents()
			If control Is Nothing Then
				Return
			End If
			RemoveHandler control.UpdateUI, AddressOf OnUpdateUI
		End Sub

		Private Sub OnUpdateUI(ByVal sender As Object, ByVal e As EventArgs)
			If button_Renamed Is Nothing Then
				Return
			End If

			Dim command As Command = CreateCommand()
			If command IsNot Nothing Then
				Dim state As New CommandButtonUIState(button_Renamed)
				command.UpdateUIState(state)
			End If
		End Sub

		Private Sub OnClick(ByVal sender As Object, ByVal e As EventArgs)
			Dim command As Command = CreateCommand()
			If command IsNot Nothing Then
				command.Execute()
			End If
		End Sub

		' You may override this method to create a command
		Protected Overridable Function CreateCommand() As Command
			Return command_Renamed
		End Function
	End Class
	#End Region ' #commandbuttonadapter
End Namespace
